### PR TITLE
bond environments: bond-atom neighbouring

### PR DESCRIPTION
**Pull request by liam-o-marsh**
the changes I talked about, split into three commits that can be reviewed separately

There's just one point where I know feedback is needed, it's for the new API for the systems.
(specifically, the huge incoherence between `pairs_per_species` and `triplets_per_species`, that can be seen in the comments of the two methods)

----
:books: Documentation preview :books:: https://rascaline--240.org.readthedocs.build/en/240/

### CODE REVIEW

**Comment by liam-o-marsh** [comment]
sure!
"BAtriplet"s are short for Bond/Atom triplets. They contain three atoms, two grouped in a bond (that acts as a center), and a third atom which acts as a neighbour.

I'm not sure what alternatives there would be to using this object explicitely.

If I were to create a new system with "ghost atoms" to serve as bond centers, I could represent the old system's triplets with the new system's pairs? But a lot of the pairs would end up being bond/bond or atom/atom, and would go to waste. Unless I deliberately break `compute_pairs` to not compute all of them.
Plus, I need the orientation of the bond, and I don't think I can fit that info in the ghost atoms.

I guess one way to bypass changing the system trait would be to have `calculators::bondatom_neighbor_list::BANeighborList` call directly `systems::neighbors::TripletNeighborsList` but I'm not sure how many things I would have to reroute and how rascaline-idiomatic it is to do so.

**Comment by Luthaf** [comment]
> I'm not sure what alternatives there would be to using this object explicitely.

Inside `SphericalExpansionForBonds` or the other calculator, you could get two neighbors list, one with the bond cutoff and one with the neighbors cutoff:

```rs
system.compute_neighbors(bond_cutoff);
bonds = system.pairs();

system.compute_neighbors(atoms_cutoff);

let mut ba_triplets = vec![];
for bond in bonds {
    let mid_point = positions[bond.first] + bond.vector / 2;

    for pair in system.pairs_containing(bond.first) {
        if distance_from_mid_point_is_below_cutoff {
            ba_triplets.push(bond, other_atom_in_the_pair);
        }
    }

    for pair in system.pairs_containing(bond.second) {
        if distance_from_mid_point_is_below_cutoff {
            ba_triplets.push(bond, other_atom_in_the_pair);
        }
    }
}
```

This way, no changes to the system are required, and you can still get the data you need.

--- 

> I guess one way to bypass changing the system trait would be to have calculators::bondatom_neighbor_list::BANeighborList call directly systems::neighbors::TripletNeighborsList but I'm not sure how many things I would have to reroute and how rascaline-idiomatic it is to do so.

This is not possible, because the `System` trait is here to abstract over system providers, meaning the system you get as input to your calculator are not necessarily `systems::SimpleSystem`. They could be something provided in `rascaline-torch`: https://github.com/Luthaf/rascaline/blob/55ad1f14363812639c865801a6d517d8c5f363f6/rascaline-torch/include/rascaline/torch/system.hpp#L19-L24

**Comment by liam-o-marsh** [comment]
oh, I see, thanks!
Now that I managed to work on this a little, I found another problem:
No matter how i slice it, the calculators I create need to have access to this list of triplets in their `.keys()`, `.samples()`, `.positions_gradient_samples()` and `.compute()`.
And for now, the only possibilities I see are really not satisfactory:

- recomputing the triplet list every time
- implementing some sort of global cache that keeps them per-system (which… would be a hashmap based on… the memory address of the systems? that doesn't seem sound. But then again good luck having a `HashMap<dyn System,_>`)
- having the calculator contain a `Vec<_>` cache whose lifetime depends (explicitely or not) on the lifetime of the `&mut [Box<dyn System>]` used with the calculator?
  The explicit dependency does not seem possible. For the implicit dependency, how do we detect that it is the same `&mut [Box<dyn System>]` passed to the different methods of the calculator?

If it is true that none of those ideas would cut it, how do you think I should proceed?

Thanks, and sorry for the late reply.

**Comment by Luthaf** [comment]
Another solution would be to allow the calculator to store data in the system, but I don't love the implications of this either.

I'll think about this over the weekend, and then maybe we can have a chat to find the best way forward!

**Comment by liam-o-marsh** [comment]
sure!
have a nice weekend, and see you then!

**Comment by liam-o-marsh** [comment]
A thing I thought about for this would be the possibility to  add other methods to BaseCalculator that would be called with a `&mut [Box<dyn System>]`, but also `caches: &mut [BTreeMap<String,Any>]`.
and the default version of these new methods (implemented in trait) would fall back to keys,samples,compute,etc...

not exactly great either (and it feels way jankier than just allowing systems to carry a cache around), but probably less disruptive than changing the System trait.

**Comment by Luthaf** [comment]
Yes, this kinda what I landed on over the weekend. Replace the input of `compute` from `&mut [Box<dyn System>]` to `&mut [SystemWithData]` (name to be improved) where 

```rs
struct SystemWithData {
    system: Box<dyn System>,
    data: BTreeMap<String,Any>
}
```

I think this is the cleanest solution here

**Comment by liam-o-marsh** [comment]
that does sound way better than what I suggested.
Thank you!

**Comment by Luthaf** [comment]
I've implemented this in the `system-with-data` branch (the struct is called `System`, the trait is now `SystemBase`). Could you try to use it and let me know if this works for you? 

For now, I allowed the data to be `Any`, but I'm also thinking to restrict it to `TensorBlock` instead (since we will have a similar mechanism in metatensor for atomistic simulations). Here as well, let me know if a `TensorBlock` would be enough for you or if you need the extra flexibility of Any!

**Comment by liam-o-marsh** [comment]
ah, should have replied earlier.
thanks, I'll try to rebase my changes on top of that, sometimes next week.
(found a thorny math bug in some of my own code, it will take a while to actually pinpoint)

**Comment by liam-o-marsh** [comment]
and that should be it
(…just note that it now has the changes from the systems_with_data branch, but I can't switch the branch of the PR)

**Comment by liam-o-marsh** [comment]
I'll be using a PR within m-stack-org to make sure it passes CI tests. (which catches more errors than cargo test and cargo bench)

File: rascaline/src/calculators/bondatom_neighbor_list.rs [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+/// if the `bond_conbtribution` option is active
+/// (When periodic boundaries arise, atom which  must not be the same may be images of each other.)
+///
+/// This sample produces a single property (`"distance"`) with three components
+/// (`"vector_direction"`) containing the x, y, and z component of the vector from
+/// the center of the triplet's 'bond' to the triplet's 'third atom', in the bond's canonical orientation.
+/// 
+/// In addition to the atom indexes, the samples also contain a pair and triplet index,
+/// to be able to distinguish between multiple triplets involving the same atoms
+/// (which can occur in periodic boundary conditions when the cutoffs are larger than the unit cell).
```
Comment (Luthaf) [inline rascaline/src/calculators/bondatom_neighbor_list.rs:189]: We changed this recently for the usual neighbor list, replacing the pair id with the cell shift for the pair (https://github.com/Luthaf/rascaline/pull/221). Could you do something similar here?

File: rascaline/src/systems/mod.rs [hunk 2]
```diff
@@ -69,6 +69,7 @@ pub trait SystemBase: Send + Sync {
     /// distance between atoms is actually bellow the cutoff passed in the last
     /// call to `compute_neighbors`. This function is only valid to call after a
     /// call to `compute_neighbors`.
+    /// TODO or CHECK: is there a risk of double counting acoss pbc?
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/systems/mod.rs:72]: Pairs are supposed to only be counted once, as `i - j` and not as `j - i`.
Is there a risk that going across boundaries (and potentially with other edge-cases at the same time) could cause a pair to be counted twice as `i - j[+1]` and `j - i[-1]` (especially if `i == j`) ?

File: rascaline/src/systems/neighbors.rs [hunk 3]
```diff
@@ -156,17 +157,17 @@ impl CellList {
...
     }
 
-    /// Add a single atom to the cell list at the given `position`. The atom is
-    /// uniquely identified by its `index`.
+    /// Add a single atom to the cell list at the given `position`.
+    /// ASSUMPTION: The atom is *uniquely* identified by its `index`.
     pub fn add_atom(&mut self, index: usize, position: Vector3D) {
         let fractional = if self.unit_cell.is_infinite() {
-            position
+            position  // TODO: this feels like it would cause bugs if we have more than one subcell (for example for a cutoff <0.5)
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/systems/neighbors.rs:170]: I need to check again, but I think there might have been an issue with small cutoffs still causing periodic boundaries? Or something? I can't remember what it was exactly.

File: rascaline/src/calculators/bondatom_neighbor_list.rs [hunk 4]
```diff
@@ -0,0 +1,909 @@
...
+/// (first and second atom), and the other neighbor is a simple atom (third atom).
+/// Both the length of the bond and the distance between neighbors are subjected to a spherical cutoff.
+/// 
+/// Unlike the corresponding pre_calculator, this calculator focuses on storing
+/// the canonical-orientation vector between bond and atom, rather than the bond vector and 'third vector'.
+///
+/// Users can request either a "full" neighbor list (including an entry for both
+/// `i - j` bonds and `j - i` bonds) or save memory/computational by only
+/// working with "half" neighbor list (only including one entry for each `i - j`
+/// bond)
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:175]: oh dear.
I thought I dealt with all the documentation. In theory it should be `i - j  +k` (which I hope is understandable)

File: rascaline/src/calculators/bondatom_neighbor_list.rs [hunk 5]
```diff
@@ -0,0 +1,909 @@
...
+/// if memory is saved, the order of i and j is that the atom with
+/// the smallest Z (or species ID in general) comes first.
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:177]: will add to next commit

File: rascaline/src/calculators/soap/spherical_expansion_bondcentered.rs [hunk 6]
```diff
@@ -0,0 +1,542 @@
...
+
+use super::super::CalculatorBase;
+
+use super::{SphericalExpansionForBondType, SphericalExpansionForBondsParameters};
+use super::spherical_expansion_pair::GradientsOptions;
+
+use super::super::{split_tensor_map_by_system, array_mut_for_system};
+
+
+/// The actual calculator used to compute SOAP spherical expansion coefficients
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/calculators/soap/spherical_expansion_bondcentered.rs:23]: Updated the docstring, removed a blank line. Not sure what's wrong with the first line of this snippet.

File: rascaline/src/pre_calculators/mod.rs [hunk 7]
```diff
@@ -0,0 +1,2 @@
+mod bond_atom_triplet_neighbors;
+pub(crate) use bond_atom_triplet_neighbors::{BATripletInfo,BATripletNeighborList};
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/pre_calculators/mod.rs:2]: Yes, part of this pre-calculator stuff is to separate what stores data in systems rather than in tensormaps.
The other part is to have a clearer dependency tree, given that part of labels/ depends on it.

File: rascaline/src/calculators/bondatom_neighbor_list.rs [hunk 8]
```diff
@@ -0,0 +1,909 @@
...
+
+use super::CalculatorBase;
+
+use crate::{Error, System};
+use crate::types::{Vector3D,Matrix3};
+use crate::pre_calculators::{BATripletInfo,BATripletNeighborList};
+
+
+
+fn rotate_vector_to_z(vec: Vector3D) -> Matrix3 {
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:16]: done

File: rascaline/src/calculators/bondatom_neighbor_list.rs (hunk 1 above)
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:189]: sure! I'll look into that

File: rascaline/src/calculators/bondatom_neighbor_list.rs [hunk 9]
```diff
@@ -0,0 +1,909 @@
...
+/// This sample produces a single property (`"distance"`) with three components
+/// (`"vector_direction"`) containing the x, y, and z component of the vector from
+/// the center of the triplet's 'bond' to the triplet's 'third atom', in the bond's canonical orientation.
+/// 
+/// In addition to the atom indexes, the samples also contain a pair and triplet index,
+/// to be able to distinguish between multiple triplets involving the same atoms
+/// (which can occur in periodic boundary conditions when the cutoffs are larger than the unit cell).
+#[derive(Debug, Clone)]
+#[derive(serde::Deserialize, serde::Serialize, schemars::JsonSchema)]
+pub struct BANeighborList {
```
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:192]: I didn't know NeighboursList was used directly.
In my case, though, I kinda need the calculations this calculator does.
Maybe I should remove this object but merge what it does in what's currently in pre_calculators?

[the remark this answers is not in the document]
Comment (Luthaf) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:192]: yes, this would be quite a bit simpler =). Same for `SphericalExpansionForBondType` and `SphericalExpansionForBond`!

[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:192]: I…
Maybe not for one of them, I still would access this spherical decomposition directly for now, so I can plug it into the rest of my L2SOAP stuff that's currently in python.
(yes, I ended up forgetting to register the right calculators to python)

File: rascaline/src/calculators/bondatom_neighbor_list.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:175]: fixed

File: rascaline/src/calculators/bondatom_neighbor_list.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:192]: Yes, this is what I mean: you are using only one of them (I guess `SphericalExpansionForBond`), so this one should be the only one for which you go through the pain and work of providing it as a calculator =)

File: rascaline/src/pre_calculators/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/pre_calculators/mod.rs:2]: should I still move the precalculator with the calculator, or does the "clearer dependency tree" argument make the new directory worth it?

[the remark this answers is not in the document]
Comment (Luthaf) [reply rascaline/src/pre_calculators/mod.rs:2]: ok, I missed that labels depends on this code. Might make sense to put it in `systems/bond_atoms_neighbors.rs` or something similar then

File: rascaline/src/systems/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply rascaline/src/systems/mod.rs:72]: Right, I'm looking into something similar right now, I'll triple check that. But pairs between an atom and it's image should only be produced once per direction (i.e. you'll get a pair between the atoms in the central image and the (1 1 0) image, but no pair between the atom in the central image and the (-1 -1 0) image.

File: rascaline/src/calculators/bondatom_neighbor_list.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:175]: So you only include one of `i-j k` and `j-i k`, but since i-j and k have different roles there is no symmetry between these.

File: rascaline/src/calculators/soap/spherical_expansion_bondcentered.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Luthaf) [reply rascaline/src/calculators/soap/spherical_expansion_bondcentered.rs:23]: I mainly meant this is very short documentation for the calculator: this should explain a bit more what is being computed, and which samples/components/properties are being produced.

File: rascaline/src/calculators/bondatom_neighbor_list.rs (hunk 1 above)
Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:189]: do I still need to do that if this is no longer exposed as a calculator?
... it probably needs to be exposed at the precalculator (or system/bond_atom_neighbors) level.

Comment (Luthaf) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:189]: you don't need to change the samples, but it might make help to make the rest of the code more understandable. You could also just use a `Vec<BATriplets>` at the pre-calculator level, and remove the need for pair/ba-triplet ids altogether

Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:189]: I though the idea of the System cache would be to mainly use TensorBlocks?
But if that's no longer a constraint, I'll switch to only using a Vec<BATriplet> rather than converting back and forth to a tensorblock

Comment (Luthaf) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:189]: I'm not sure. For now the code in rascaline accepts everything. There is something similar happening as part of https://github.com/lab-cosmo/metatensor/pull/405, where metatensor is getting a `System` definition to define, export and run MLIP; and there the data is only TensorBlock.

If having `Any` here makes your life easier, I would stick with it, since the goals of the two API are a bit different.

Comment (liam-o-marsh) [reply rascaline/src/calculators/bondatom_neighbor_list.rs:189]: ah OK, thanks!

**Comment by liam-o-marsh** [comment]
and most of your remarks should be taken into account (notable exception: the ones that were about "TODO"s left in the code)

File: rascaline/src/systems/neighbors.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/systems/neighbors.rs:170]: in the end, there are no bugs here, eventhough this is far from the best, performance-wise.

File: rascaline/src/systems/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/systems/mod.rs:75]: I just wrote this test in systems/neighbors.rs, and it fails:
```rust
    #[test]
    fn small_cell_large_cutoffs() {
        let cell = UnitCell::cubic(0.5);
        let positions = [Vector3D::new(0.0, 0.0, 0.0)];
        let neighbors = NeighborsList::new(&positions, cell, 0.6);

        let expected = [
            (Vector3D::new(0.0, 0.0, 0.5), [0, 0, 1]),
            (Vector3D::new(0.0, 0.5, 0.0),  [0, 1, 0]),
            (Vector3D::new(0.5, 0.0, 0.0),  [1, 0, 0]),
        ];
        
        assert_eq!(neighbors.pairs.len(), 3);
        for (pair, (vector, shifts)) in neighbors.pairs.iter().zip(&expected) {
            assert_eq!(pair.first, 0);
            assert_eq!(pair.second, 0);
            assert_ulps_eq!(pair.distance, 0.5);
            assert_ulps_eq!(pair.vector, vector);
            assert_eq!(&pair.cell_shift_indices, shifts);
        }
    }
 ```
and this one (which I think is the wrong result) succeeds:
```rust
    #[test]
    fn small_cell_large_cutoffs() {
        let cell = UnitCell::cubic(0.5);
        let positions = [Vector3D::new(0.0, 0.0, 0.0)];
        let neighbors = NeighborsList::new(&positions, cell, 0.6);

        let expected = [
            (Vector3D::new(-0.5, 0.0, 0.0),  [-1, 0, 0]),
            (Vector3D::new(0.0, -0.5, 0.0),  [0, -1, 0]),
            (Vector3D::new(0.0, 0.0, -0.5), [0, 0, -1]),
            (Vector3D::new(0.0, 0.0, 0.5), [0, 0, 1]),
            (Vector3D::new(0.0, 0.5, 0.0),  [0, 1, 0]),
            (Vector3D::new(0.5, 0.0, 0.0),  [1, 0, 0]),
        ];
        
        assert_eq!(neighbors.pairs.len(), 6);
        for (pair, (vector, shifts)) in neighbors.pairs.iter().zip(&expected) {
            assert_eq!(pair.first, 0);
            assert_eq!(pair.second, 0);
            assert_ulps_eq!(pair.distance, 0.5);
            assert_ulps_eq!(pair.vector, vector);
            assert_eq!(&pair.cell_shift_indices, shifts);
        }
    }
```

[the remark this answers is not in the document]
Comment (liam-o-marsh) [reply rascaline/src/systems/mod.rs:75]: if the behaviour for the first one is the correct one, then the latest commit (https://github.com/Luthaf/rascaline/pull/240/commits/4c6354a41531a3b0820ca9e489edf908d2402ab0) fixes this (it can be cherry-picked as well)

[the remark this answers is not in the document]
Comment (Luthaf) [reply rascaline/src/systems/mod.rs:75]: ok, that does look like a bug! Thanks for looking into it =) I don't have much time right now to fully evaluate the consequences, but I'll cherry-pick your commit asap!

**Comment by liam-o-marsh** [comment]
question:
are there specific things I need to do here before it gets merged?

**Comment by Luthaf** [comment]
Sorry for the delay, I was deep in another part of development which is now reaching its end; so I should have more time for rascaline work.

There is not much you need to do for now, I have to do another full round of review on this & finish + merge #260.

----

One point I wanted to discuss with you was how we could expose the fact that this is a very new, still experimental representation that might be changed if we realise it needs to do some things differently. My idea here was to have the code in the main library and make sure it continues working & passing tests, but force the user to opt-in and acknowledge this is experimental code if they want to use it. 

An environment variable would be a good way to do this:

```bash
export RASCALINE_EXPERIMENTAL_BOND_ATOM_SPX=1
python script.py
```

or 

```
import os
os.environ["RASCALINE_EXPERIMENTAL_BOND_ATOM_SPX"] = "1"

from rascaline import BondAtomSphericalExpansion
```

What do you think? 

Another point about this is when will the code move out of this experimental state and always be available. I'm thinking having one published paper with it + the main author (you in this case, but we will use similar mechanism for other new representation later) agreeing that the code is ready for general use.

File: python/rascaline/rascaline/systems/__init__.py [hunk 10]
```diff
@@ -32,6 +33,15 @@ def __init__(self):
     .. _chemfiles.Frame: http://chemfiles.org/chemfiles.py/latest/reference/frame.html
     """
 
+if HAVE_PYSCF:
```
Comment (Luthaf) [inline python/rascaline/rascaline/systems/__init__.py:36]: Could you make the PYSCF system into a separate PR?

Comment (liam-o-marsh) [reply python/rascaline/rascaline/systems/__init__.py:36]: ...bisecting only found the commit where equistore was renamed to metatensor, breaking the build on all older rascaline commits, due to a failed hash check

**Comment by liam-o-marsh** [comment]
> Sorry for the delay, I was deep in another part of development which is now reaching its end; so I should have more time for rascaline work.
> There is not much you need to do for now, I have to do another full round of review on this & finish + merge #260.
OK! To be honest, I was also busy, with my candidacy exam. 

> ```shell
> export RASCALINE_EXPERIMENTAL_BOND_ATOM_SPX=1
> ```
Sure!

> Another point about this is when will the code move out of this experimental state and always be available. I'm thinking having one published paper with it + the main author (you in this case, but we will use similar mechanism for other new representation later) agreeing that the code is ready for general use.

That seems like a good rule of thumb for this sort of thing. Thank you for thinking about this!

File: python/rascaline/rascaline/systems/__init__.py (hunk 10 above)
Comment (Luthaf) [reply python/rascaline/rascaline/systems/__init__.py:36]: > valgrind is annoyed at entries 5 and 9 of cargo test -p rascaline-c-api --test run-cxx-tests
(a problem that pre-dates my changes)

Hmm, can you open a separate issue for this? Or just share the setup with me on Slack? We run the valgrind tests on CI, so this is strange.

Comment (liam-o-marsh) [reply python/rascaline/rascaline/systems/__init__.py:36]: well nevermind the error disappears for a fresh clone of master

Comment (liam-o-marsh) [reply python/rascaline/rascaline/systems/__init__.py:36]: I'll need to look more into this, but when it does/doesn't happen seems to be nonsense for now.

**Comment by liam-o-marsh** [comment]
oh no
er.... I might have broken the PR...
I forgot it targeted the system-with-data branch rather than the main branch...
(speaking of, isn't system-with-data ready for prime time by now?)

**Comment by Luthaf** [comment]
>  speaking of, isn't system-with-data ready for prime time by now?

Yes, it should be. I mainly need to document and merge it!

**Comment by liam-o-marsh** [comment]
except I probably need to rebase this further because of the name change, which seems to break some tests
